src/protocol/libp2p/kademlia/query/find_node.rs

```diff
@@ -117,120 +122,373 @@ impl<T: Clone + Into<Vec<u8>>> FindNodeContext<T> {
         // calculate distance for `peer` from target and insert it if
         //  a) the map doesn't have 20 responses
         //  b) it can replace some other peer that has a higher distance
-        let distance = self.target.distance(&peer.key);
+        let distance = self.config.target.distance(&peer.key);
 
         // always mark the peer as queried to prevent it getting queried again
         self.queried.insert(peer.peer);
 
-        // TODO: could this be written in another way?
-        // TODO: only insert nodes from whom a response was received
```

Comment:
Is this TODO still relevant? It looks like nothing has changed in its respect.

Reply:
I dont think this is addressed indeed, will briefly look over libp2p or create an issue if this may be too involved / not covered by our testing atm, thanks!

Reply:
While looking at this I realized a few things:
- we were registering the first 20 (repl factor) responses
- whenever we got a new response (better) we tried to update just the _peerID__ of the __last entry__

This has a few implications:
- the first 19 entries are never updated with a better (closer) peer
- the distance of the last entry was never updated. And because of this, we might end up in a situation where the record is updated with a furthest distance

```rust
// Presume: 
last_entry: distance = 10, peer = A

// Registering a new response
new_response: distance = 2, peer = B
-> last entry: distance = 10, peer = B

// Registering a new response
new_response: distance = 7, peer = C
-> last entry: distance = 10, peer = C

// we should've kept B as best
```

In this case, we should have updated other records and only keep the window of the best 20 peers.

To mitigate this, if we receive a response better than the last entry: 
- insert the response with distance and peer to our btreemap, if and only if we didn't track that distance before
- pop the last entry to keep only K entries

Have also added a test for this `keep_k_best_results`